yeaphone.1
.\"                              hey, Emacs:   -*- nroff -*-
.\" yeaphone is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; see the file COPYING.  If not, write to
.\" the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH YEAPHONE 1 "January 27, 2009"
.\" Please update the above date whenever this man page is modified.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins (default)
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
yeaphone \- A SIP phone for Yealink USB-P1K/P1KH/P4K handsets using liblinphone
.SH SYNOPSIS
.B yeaphone
.RI [ options ]
.SH DESCRIPTION
The goal of the \fBYeaphone\fP project is to provide a user interface on a
Yealink USB handset <http://www.yealink.com/en/index.asp> for the VoIP
software Linphone thus making a PC keyboard and monitor unneccessary.
This makes Yeaphone ideal for "Embedded Devices" which work very energy
efficient but typically need an extra devices for user interaction (in
this case the handset).
.PP
Its features include
    * Display the phone number of incoming and outgoing calls
    * Show the duration of a call
    * Allows to store and recall 10 phone numbers
    * Mute the ring tone of incoming calls
    * Adjustable volume of the speaker
.PP
Yeaphone has to be configured through linphonec, the console version
of Linphone. The following example shows how to get a working connection
to a VoIP provider.
.PP
    root@NSLU2:~# \fBlinphonec\fP
    Ready
    linphonec> \fBproxy add\fP
    Adding new proxy setup. Hit ^D to abort.
    Enter proxy sip address: \fBsip:sipgate.at\fP
    Your identity for this proxy: \fBsip:1234567@sipgate.at\fP
    Do you want to register on this proxy (yes/no): \fByes\fP
    Specify register expiration time in seconds (default is 600): \fB600\fP
    Expiration: 600 seconds
    Specify route if needed: \fB[leave empty]\fP
    No route specified.
    --------------------------------------------
    sip address: sip:sipgate.at
    route:
    identity: sip:1234567@sipgate.at
    register: yes
    expires: 600
    --------------------------------------------
    Accept the above proxy configuration (yes/no) ?: \fByes\fP
    Proxy added.
    linphonec>
    Password for 1234567 on "sipgate.at": \fB[enter password]\fP
    Registration on sip:sipgate.at sucessful.
    linphonec>

Using linphonec the VoIP connection can be tested immediately before
using Yeaphone. For the speaker and microphone the default audio card
is used, and 
commands still have to be entered through the PC keyboard
though. The most important commands of linphonec are:

    * help [<command>]
      shows the available commands and provides a detailed description
      for each command
    * call <phone number or SIP address>
      dials the specified phone number of SIP address
    * answer
      answers an incoming call
    * terminate
      terminates an (incoming or outgoing) call
    * quit
      unregisters at the VoIP provider and quits linphonec

As soon as calls can be made this way, linphonec can be stopped and
yeaphone can be invoked. Its output should look similar to this:

    root@NSLU2:~# \fByeaphone\fP
    path_sysfs = /sys/bus/usb/drivers/yealink/3-1.3:1.3/
    path_buf = /dev/input/event2
    Detected handset Yealink USB-P1K
    Ready
    Registration on sip:sipgate.at sucessful.

At the same time the current date and time are show on the handset's
display, a quick sequence of status messages appear, and finally the
lower part of the display should be empty. Now the handset's keyboard
can be used to place and receive calls.

The various keys of a P1K/P1KH have the following functions:

   1. before a phone call
          * 0-9,*,#
            Dial a phone number
          * C
            Remove the last character of the dialed number
          * C pressed long
            Remove the complete number
          * <red key>
            Remove the complete number
          * <green key>
            Initiate the call
          * <up key> + <0-9>
            No number dialed yet: Recall and display a number from
            memory X
            Number already dialed: Store the currently displayed number
            at memory X
   2. during a phone call
          * 0-9,*,#
            Generate DTMF tones
          * <red key>
            Terminate the call
          * VOL+/-
            Adjust the speaker's volume 
   3. after a phone call (with the last number displayed)
          * 0-9,*,#
            Dial a new phone number
          * C
            Remove the complete number
          * <red key>
            Remove the complete number
          * <green key>
            Redial the last number
          * <up key> + <0-9>
            Store the last number to memory /X/ 
   4. when the phone rings
          * #
            Mute the ringing without picking up
          * <red key>
            Immediately terminate the call without picking up
          * <green key>
            Pick up 
   5. general
          * <red key> pressed long
            Start and stop the VoIP connection (corresponds to turning
            on/off a phone) 

For a P4K (desk phone) the keys have the same meaning, except that starting and
finishing a phone call is done by picking up or hanging up the handset.

To correctly understand the phone number of incoming calls, Yeaphone
needs some adjustments in the file \fB~/.yeaphonerc\fP. After terminating
yeaphone the parameters \fBintl-access-code\fP, \fBnatl-access-code\fP, and
\fBcountry-code\fP should be changed accordingly, the default values work for
Austria only.

In \fB~/.yeaphonerc\fP you can also spedify custom ringtones (P1K/P1KH only)
for different numbers by adding lines according to the following example:
  ringtone_default   /usr/share/yeaphone/ringtones/default_p1k.bin
  ringtone_01234567  /usr/share/yeaphone/ringtones/special_p1k.bin
  ringtone_0555777   doorbell_p1k.bin

If you specify relative paths to the ringtones, they are based on
$HOME/.yeaphone/ringtone.

Another feature to be configured in \fB~/.yeaphonerc\fP is the minimum ring
duration. If for a certain caller ID the duration of the ring should be at
least 5 seconds, this can be specified as:
  minring_01234567  5

If the ringtone of a P4K should be sent to a different audio device, the name of this
device (preceeded by "ALSA: ") can
be specified by the option below. Note that the wav-file to be played is still
set in \fB~/.linphonerc\fP.
  ringer-device  ALSA: USB Audio

.SH OPTIONS
.TP
\fI\-\-id=<id>\fP
Attach to the device with an ID <id>.
.TP
\fI\-w, \-\-wait=[<sec>]\fP
Check for the handset every <sec> seconds (default: 10s)
.TP
\fI\-v, \-\-verbose\fP
Show debug messages.
.TP
\fI\-h, \-\-help\fP
Print this help message.

.SH FILES
.I ~/.linphonerc
.RS
Linphone's configuration file which can be created using \fBlinphonec\fP
(see above).
.RE
.I ~/.yeaphonerc
.RS
Yeaphone's additional configuration file which until now has to be edited manually.
It contains the stored numbers and localization codes.
.RE
.SH AUTHOR
Thomas Reitmayr <treitmayr@devbase.at>.

.SH "SEE ALSO"
.BR linphonec (1)


ChangeLog
2006-09-24  Thomas Reitmayr  <treitmayr@devbase.at>

	* yeaphone: initial version.

2008-07-09  Thomas Reitmayr  <treitmayr@devbase.at>

  * src/yldisp.c:
    Added strategies for finding the event device which also work with newer
    kernel versions.



src/yeaphone.c
/****************************************************************************
 *
 *  File: yeaphone.c
 *
 *  Copyright (C) 2006 - 2008  Thomas Reitmayr <treitmayr@devbase.at>
 *
 ****************************************************************************
 *
 *  This file is part of Yeaphone.
 *
 *  Yeaphone is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Library General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 *
 ****************************************************************************/

#include <mcheck.h>

#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include <signal.h>
#include "ylsysfs.h"
#include "yldisp.h"
#include "lpcontrol.h"
#include "ylcontrol.h"
#include "ypconfig.h"
#include "ypmainloop.h"

#ifdef DMALLOC
#include <dmalloc.h>
#endif


#define CONFIG_FILE ".yeaphonerc"

char *mycode = "43";   /* default to Austria ;) */
int ylcontrol_started = 0;


struct cmdline_options {
  char *uniq;
  int wait_for_device;
  int verbose;
};
static struct cmdline_options cmdline_opts = {
  uniq: NULL,
  wait_for_device: 0,
  verbose: 0
};

void parse_args(int argc, char **argv) {
  int c;
  int opt_index;
  static struct option long_options[] = {
    {"help", 0, 0, 'h'},
    {"id", 1, 0, 0},
    {"wait", 2, 0, 1},
    {"verbose", 0, 0, 'v'},
    {0, 0, 0, 0}
  };

  while ((c = getopt_long(argc, argv, "vhw", long_options, &opt_index)) >= 0) {
    switch (c) {
    case 0:
      cmdline_opts.uniq = strdup(optarg);
      break;
    case 1: 
      cmdline_opts.wait_for_device = (optarg) ? atoi(optarg) : 10;
      break;
    case 'w': 
      cmdline_opts.wait_for_device = 10;
      break;
    case 'v':
      cmdline_opts.verbose = 1;
      break;
    default:
      printf("Usage: yeaphone [options]\n");
      printf("\t--id=<id>\tAttach to the device with an ID <id>.\n");
      printf("\t--wait=[<sec>]\tCheck for the handset every <sec> seconds.\n");
      printf("\t-w\t\tCheck for the handset every 10 seconds.\n");
      printf("\t--verbose|-v\tShow debug messages.\n");
      printf("\t--help|-h\tPrint this help message.\n");
      exit(1);
    }
  }
}


void read_config() {
  char *cfgfile;
  char *home;
  int len = 0;
  
  home = getenv("HOME");
  if (home) {
    len = strlen(home) + strlen(CONFIG_FILE) + 2;
    cfgfile = malloc(len);
    strcpy(cfgfile, home);
    strcat(cfgfile, "/"CONFIG_FILE);
  }
  else {
    cfgfile = strdup(CONFIG_FILE);
  }
  
  ypconfig_read(cfgfile);
  
  free(cfgfile);
}


void terminate(int signum)
{
  if (ylcontrol_started) {
    puts("\ngraceful exit requested...");
    stop_ylcontrol();
  }
  else {
    signal(signum, SIG_DFL);
    raise(signum);
  }
}


int main(int argc, char **argv) {
  int ret;

  parse_args(argc, argv);
  read_config();
  
  /* graceful exit handler */
  signal(SIGTERM, &terminate);
  signal(SIGINT, &terminate);
  
  yp_ml_init();
  init_ylcontrol(mycode);
  ylcontrol_started = 0;

  while (1) {
    ret = ylsysfs_find_device(cmdline_opts.uniq);
    if (ret == -ENOENT) {
      if (cmdline_opts.wait_for_device) {
        printf("Please connect your handset, waiting...\n");
        sleep(cmdline_opts.wait_for_device);
        continue;
      }
      else {
        printf("No appropriate handset found, exiting...\n");
        break;
      }
    }
    else if (ret < 0) {
      break;
    }

    start_lpcontrol(1, NULL);
    start_ylcontrol();
    ylcontrol_started = 1;

    if (yp_ml_run() != 0)
      break;

    ylcontrol_started = 0;
    yldisp_clear();
  }

  return 0;
}

